lib/assets/chart_builder/main.js

```diff
@@ -1,358 +1,306 @@
+                v-model="fields.color_field"
+                :options="axisOptions"
+                :count
+                :disabled="noDataVariable"
+              />
+              <BaseSelect
+                name="color_field_type"
+                label="Type"
+                v-model="fields.color_field_type"
+                :options="typeOptions"
+                :disabled="noColorField"
+              />
+              <BaseSelect
+                name="color_field_aggregate"
+                label="Aggregate"
+                v-model="fields.color_field_aggregate"
+                :options="aggregateOptions"
+                :disabled="noColorField"
+              />
             </div>
           </div>
         </div>
+        </form>
       </div>
-    </div>
-  `;
-
-  const channelFields = ["x_field", "y_field", "color_field"];
-  const typeFields = ["x_field_type", "y_field_type", "color_field_type"];
-  const aggregateFields = ["x_field_aggregate", "y_field_aggregate", "color_field_aggregate"];
-
-  const state = {
-    dataOptions: payload.data_options,
-  };
-
-  const selectDataElem = ctx.root.querySelector(`[name="data_variable"]`);
-
-  updateInfoBox(payload.missing_dep);
-
-  if (payload.fresh) {
-    setFreshData(payload.data_options);
-  } else {
-    renderCurrentData(payload.fields, payload.data_options);
-  }
-
-  ctx.root.addEventListener("change", handleFieldChange);
-  selectDataElem.addEventListener("change", updateAxisOptions);
-  channelFieldsListener();
-
-  function handleFieldChange(event) {
-    const { name, value } = event.target;
-    ctx.pushEvent("update_field", { field: name, value });
-  }
-
-  function updateAxisOptions(event) {
-    const { value } = event.target;
-    const axisOptions = state.dataOptions[value];
-    setAxisOptions(axisOptions);
-  }
-
-  function maybeResetFieldOptions(event) {
-    const name = event.target.name;
-    const selectElemType = ctx.root.querySelector(`[name="${name}_type"]`);
-    const selectElemAggregate = ctx.root.querySelector(`[name="${name}_aggregate"]`);
-    const { value } = event.target;
-    if (value && value !== "__count__") {
-      selectElemType.disabled = false;
-      selectElemAggregate.disabled = false;
-    } else {
-      resetField(selectElemType);
-      resetField(selectElemAggregate);
+    `,
+
+    data() {
+      return {
+        fields: payload.fields,
+        dataOptions: payload.data_options,
+        missingDep: payload.missing_dep,
+        chartOptions: ["bar", "line", "point", "area"],
+        typeOptions: ["quantitative", "nominal", "ordinal"],
+        aggregateOptions: ["sum", "mean"],
+        dataVariables: Object.keys(payload.data_options),
+      };
+    },
+
+    mounted() {
+      if (payload.fresh && this.dataVariables.length) {
+        this.fields.data_variable = Object.keys(this.dataOptions)[0];
+        ctx.pushEvent("update_field", { field: "data_variable", value: this.fields.data_variable });
```

Comment:
Oh, we should have this logic on the server! When we scan binding and discover data variables, but the current `data_variable` is `nil`, we should set it to the first one and broadcast together with the new variables. Having that you probably no longer need the `fresh` check :)

Reply:
> and the scan binding doing too much

Not that the check is expensive, but we would do it in `handle_info`, so it's fine either way

I don't think it's that complicated, we already do something analogous here:

https://github.com/livebook-dev/kino/blob/c6194c743299ada79d94ff12a0fd678a0aa53dfd/lib/kino/smart_cell/sql.ex#L71-L74

(connections -> data variables, connection -> current data variable)

Reply:
Btw, my primary concern with the current version is that it doesn't feel right to send a field update to the server when the client just `mount`s :)

Reply:
Which check is expensive?
Well, in this case we are actually doing an update for the user. It’s in fact a field update! If we change it to the server side, we will need to broadcast it anyway. Vue needs to know that’s a change to trigger everything correctly.
Do you have some code suggestions u think solves it?
My main concern is about the unavailable axis options. Just broadcast the set_available_data and the updated fields won’t work.
I guess we will need a specific broadcast for this case to force vue to trigger everything manually.

Reply:
Why can't we send available data together with the selected data variable and do:

```
...
app.dataOptions = data_options;
app.fields.data_variable = data_variable;
```

That's what would happen if we received a regular field update, no?